### PR TITLE
[MIG] account_global_discount: 16.0

### PR DESCRIPTION
**Pull request by ferran-S73**
Depends on:
- [ ] https://github.com/OCA/server-backend/pull/221

~~Hi I'm gonna need some help migrating this module @OCA/accounting-maintainers @kirca @omar7r @pedrobaeza~~

~~[Many accounting code has been refactored in v16](https://github.com/odoo/odoo/commit/d8d47f9ff8554f4b39487fd2f13c153c7d6f958d) and so my approach here, since there's no `_recompute_tax_lines` nor `_recompute_dynamic_lines` is to do this module's calculations whenever an invoice is created or the field `global_discount_ids` is written following indications from Odoo's commit description.~~

~~I'm not quite getting the desired results and I was wondering if you could help me pointing if this is a viable approach or if you have any other idea. 
I also looked at [this migration](https://github.com/OCA/account-invoicing/pull/1269/commits/e628d2a778179d36636af6b40c35a10b93159bc4) but can't seem to figure how could I do something similar here.~~

~~Thanks in advance to anyone who can help~~

In order to adapt this module to the new way of handling invoices in 16.0 I moved the creation / update of discount lines to the `create` and `write` methods of the `account.move`. We no longer need the recalc the bas amount for the discount lines since it is done automatically by Odoo as well as the total amounts of the invoice.

Also, for the field `global_discount_item` which was needed because `invoice_global_discount_id` wasn't properly filled, I checked and that last field is now handled correctly. I'm not removing the `global_discount_item` altogether for compatibility reasons with old invoices since there's no clear way to migrate it to fill `invoice_global_discount_id`.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I can't say too much. Maybe you have to override `_inverse_tax_totals`.

**Comment by ferran-S73** [comment]
Hey @OCA/accounting-maintainers this should be ready for review now! :partying_face:

**Comment by ferran-S73** [comment]
Hi @OCA/accounting-maintainers I'm having trouble migrating the tests for this module, for some reason when a global discount is added a new tax line is added too instead of updating the existing one. Does anyone have encountered a similar problem? I'd be very thankful if anyone could help me out here. Thanks in advance

**Comment by BT-srieskamp** [comment]
@ferran-S73 not sure if this is the right place for feedback to that module. I tested everything and it looks quite good from a users perspective. But the two new lines at the bottom of the Sale Order and Invoice are not in the right place. Maybe this is already fixed? Can you reproduce this? I added an attachment.
Screenshot 2023-04-25 at 14 01 45

**Comment by ferran-S73** [comment]
> @ferran-S73 not sure if this is the right place for feedback to that module. I tested everything and it looks quite good from a users perspective. But the two new lines at the bottom of the Sale Order and Invoice are not in the right place. Maybe this is already fixed? Can you reproduce this? I added an attachment. Screenshot 2023-04-25 at 14 01 45

Hi I'll try to reproduce with the same currency and reach you back with anything I find out @BT-skettler

**Comment by ferran-S73** [comment]
@OCA/accounting-maintainers

**Comment by rafaelbn** [comment]
Hello @ferran-S73 ,

Is this whay you except?

imagen


I have beed testing and I don't like to much this module, complicated. REAMDE in configuration should be reviewed.

Any case, testing is OK no error, just some maybe UX?

Up to you!

**Review by rafaelbn — APPROVED** [review]
Functional basic test in runbot 👍🏼

**Comment by rafaelbn** [comment]
You can review here @BT-skettler ! 🥳

**Comment by BT-srieskamp** [comment]
> 

@ferran-S73 I hope I tested it correctly. On runbot I could see that the summary below the sale order line is aligned (with currency CHF).
Thanks a lot for you effort!

**Comment by victoralmau** [comment]
Please, cherry-pick https://github.com/OCA/account-invoicing/pull/1531 to commit history

**Comment by ferran-S73** [comment]
> Please, cherry-pick #1531 to commit history

@victoralmau done

**Review by QuocDuong1306 — APPROVED** [review]
Tested, lgtm

**Review by cyrilmanuel — APPROVED** [review]
not an expert but lgtm

**Comment by rafaelbn** [comment]
Thank you @ferran-S73 ! @victoralmau your comments are attended, are you agree? 😄

**Review by leemannd — APPROVED** [review]

**Comment by linuxivan** [comment]
There is an error when returning an invoice, it re-applies the discount on the already discounted invoice and the amount is not the same, resulting in a partial payment on the original invoice.
image

**Review by JavierIniesta — APPROVED** [review]
Code Review: LGTM

Could you add unit tests to cover the missing use cases?

**Comment by linuxivan** [comment]
> Code Review: LGTM
> 
> Could you add unit tests to cover the missing use cases?

You have to look better.

**Comment by hildickethan** [comment]
> There is an error when returning an invoice, it re-applies the discount on the already discounted invoice and the amount is not the same, resulting in a partial payment on the original invoice.

Fixed, now when creating a refund it will clear the global discount line and recalc it
Also fixed a bug with the "apply discount" field not working correctly

**Review by cyrilmanuel — APPROVED** [review]
not an expert but lgtm

**Comment by simahawk** [comment]
FTR still blocked by its dependency https://github.com/OCA/server-backend/pull/221

**Comment by PauBForgeFlow** [comment]
Hi, I think it would be interesting to cherry-pick this [#PR1448 ](https://github.com/OCA/account-invoicing/pull/1448) to commit history.

**Comment by miguel-S73** [comment]
@PauBForgeFlow `cherry-pick` added.

**Review by cyrilmanuel — APPROVED** [review]

**Review by carolinafernandez-tecnativa — APPROVED** [review]
Functional review LGTM :) Thanks!
TT45863

**Comment by rafaelbn** [comment]
- [ ] Depends in https://github.com/OCA/server-backend/pull/221

File: test-requirements.txt
Comment (GasparJarsa) [inline test-requirements.txt:1]: This module is already merged, could you please remove this file in order to pass the tests in Github Actions.

**Comment by carolinafernandez-tecnativa** [comment]
Is there anything else to do? so we can unlock some migrations!
Thanks!

ping @pedrobaeza

**Comment by pedrobaeza** [comment]
For unlocking the situation, I have extracted the migration from this PR and remove the depending change in #1724 

Closing this one. When the other PR gets unlocked, you can propose the improvement in this module in a new PR.

**Pull request closed without merging**
